### ISSUE TITLE
PyCapnp, a bit abandoned but very much awesome!

### ISSUE THREAD

**Issue by Cheaterman** [capnproto/pycapnp#167]
Hi,

First I'd like to say that I've been having lots of fun lately learning how to use pycapnp. I implemented a basic chat system (https://github.com/Cheaterman/capnchat) and it works and is cool.

However the main reason I was studying capnproto is to make a networking stack for multiplayer games. As far as I can see, there's no way to get pycapnp working with UDP at this moment (capnp itself only added UDP to KJ in 0.6, I suppose these functions aren't even wrapped, so not implemented in `TwoPartyServer`) - at least not for more than a single client (because of the way passing sockets is implemented - it apparently used to apply to connection strings too).

So, considering it doesn't have what I need, and the project itself seems very much abandoned, should I give up on pycapnp and use gRPC/protobuf instead?

Thanks!

PS: FWIW, I could also make my game protocol use TCP. I'm mostly worried about this project being abandoned and my protocol having a hard dependency on it.

**Comment by Cheaterman** [comment]
I believe I can close this issue - but if others want to rant about the current lack of maintainer on this project (I believe the author himself said he doesn't have any active project using pycapnp any longer and therefore has little time to spend on it), well, I guess this is the right place. ^^"

**Comment by cjermain** [comment]
@Cheaterman, I am working to get pycapnp up to date. The previous maintainer has indeed moved on. We are looking for a long term maintainer for the project, as I only have the bandwidth for a short-term commitment. I encourage anyone landing on this issue to speak up and get involved!

**Comment by andreycizov** [comment]
I am currently working on a bunch of fixes to the library as I'll be using it in the near future.

Features done:
  - get rid of Xxx.new_message() in favour of Xxx() constructors [1].
  - Union can be constructed using their constructors [2].
  - Clearer exceptions thrown for wrong arguments, etc.

PS: 1 and 2 don't actually involve any more copying as you'd be doing by using the older, less ergonomic API.

I hope I can get it to work with asyncio so that the whole kj::Future thing could be reused with other async Python primitives. I don't think it should take a lot of work especially for the minimal feature set.

https://github.com/andreycizov/pycapnp

**Comment by Cheaterman** [comment]
That is incredible my friend! Good job!!! Since you're breaking the API anyway (and it feels like a pretty important feature for many projects), you might wanna consider merging https://github.com/capnproto/pycapnp/pull/169 into your fork :-)

I think I will be porting my code to your fork in order to give you relevant feedback. I already love the fact you're trying to make the abstraction less leaky :-)

**Comment by andreycizov** [comment]
@Cheaterman to be fair, right now the only concrete incompatibility with the older versions would be the parameter exceptions thrown whenever a method is called (e.g. wrong type/name). That's not (usually) something you're supposed to be catching anyway.

The newer constructors simply extend the types so you can use them the old way too.

Feel free to give me directions. I am currently going through the code base mainly to understand how it works under the hood.

I'll merge your commit a bit later! Otherwise I am still curious if there are many who are still using this project.

**Comment by cjermain** [comment]
@andreycizov, thanks for leading these updates. I will review your PRs and give feedback. I encourage everyone to get involved in the review process.

This project is currently looking for a long-term maintainer. I will try to keep things moving until that happens.

**Comment by haata** [comment]
@andreycizov I'm also interested in the async Python primitives and I don't mind helping out (specifically, I'm trying to get a Python Client <-SSL-> Rust Server working and I don't mind working on patches to make this happen).

**Comment by haata** [comment]
I've probably broken a bunch of rules and such, but I've managed to get very very basic asyncio remote wait working alongside an asyncio socket to interface with a TwoWayPipe (instead of file handles) for the TwoPartyClient (using thread_client.py). So far I'm only interested in client code for capnproto so I haven't really touched the server.

I'm not sure my approach is worthy of a PR, but I'll clean it up over the next few days and submit it.

**Comment by Cheaterman** [comment]
This sounds extremely promising! Does the TwoWayPipe change allow for UDP
transport, or would simplify it?

Le lun. 16 sept. 2019 à 06:23, Jacob Alexander <notifications@github.com> a
écrit :

**Comment by haata** [comment]
I'm not sure, but I'll take a look at how UDP has been done in the past using capnproto.
But as long as you set up an async reader and writer you should be able to use pretty much any container format.

Client Connection -> https://docs.python.org/3/library/asyncio-stream.html#asyncio.open_connection
Server Connection -> https://docs.python.org/3/library/asyncio-stream.html#asyncio.start_server

I was also able to get TwoPartyServer working as well (nearly all of the work had already been done for the client).

**Comment by haata** [comment]
@Cheaterman I did a bit of investigation into UDP this morning.

Option 1: Use the capnproto bindDatagramPort
- This will work, but I don't really recommend it as it'll be a bunch of work to wrap enough functions and classes to make UDP useful

Option 2: Re-assemble UDP packets into a stream and feed into the TwoWayPipe using reader and writer streams.
- This will allow you to do all the networking code in Python, not having to worry about any missing wrappers on the pycapnp side. The downside is that you'll need to handle the stream yourself (though I don't think capnproto/kj is doing anything special other than looking for truncated packets)
- In theory you only need my asyncio changes to make this work (I'm still pretty new to python asyncio though).

Also I posted my PR for asyncio https://github.com/capnproto/pycapnp/pull/190
